milli/src/documents/builder.rs

```diff
@@ -0,0 +1,64 @@
+use std::io;
+
+use bimap::BiHashMap;
+use byteorder::{BigEndian, WriteBytesExt};
+use serde::ser::Serialize;
+
+use super::serde::DocumentsSerilializer;
+use super::{ByteCounter, DocumentsMetadata, Error};
+use crate::FieldId;
+
+pub struct DocumentsBuilder<W> {
+    serializer: DocumentsSerilializer<W>,
+}
+
+impl<W: io::Write + io::Seek> DocumentsBuilder<W> {
+    pub fn new(writer: W, index: BiHashMap<FieldId, String>) -> Result<Self, Error> {
+        let mut writer = ByteCounter::new(writer);
+        // add space to write the offset of the metadata at the end of the writer
+        writer.write_u64::<BigEndian>(0)?;
+
+        let serializer =
+            DocumentsSerilializer { writer, buffer: Vec::new(), index, count: 0, allow_seq: true };
+
+        Ok(Self { serializer })
+    }
+
+    /// Returns the number of documents that have been written to the builder.
+    pub fn len(&self) -> usize {
+        self.serializer.count
+    }
+
+    /// This method must be called after the document addition is terminated. It will put the
+    /// metadata at the end of the file, and write the metadata offset at the beginning on the
+    /// file.
+    pub fn finish(self) -> Result<(), Error> {
+        let DocumentsSerilializer {
+            writer: ByteCounter { mut writer, count: offset },
+            index,
+            count,
+            ..
+        } = self.serializer;
+
+        let meta = DocumentsMetadata { count, index };
+
+        bincode::serialize_into(&mut writer, &meta)?;
+
+        writer.seek(io::SeekFrom::Start(0))?;
+        writer.write_u64::<BigEndian>(offset as u64)?;
+
+        writer.flush()?;
+
+        Ok(())
+    }
+
```

Comment:
I would like that you introduce [an `into_inner` method](https://docs.rs/fst/0.4.7/fst/struct.MapBuilder.html#method.into_inner) that consumes `self` and call `finish` before returning the inner writer. It is a better way to use this kind of wrapper type, the user gives the writer to the `DocumentsBuilder` and when you are done, you ask the wrapper type to give you back your writer, this way you can't forget to call finish.

Reply:
Well, in real usecases you don't need into_inner, you will be writing the documents to a file, and later on, re-open the file and read from it, so for real use cases, it would be even worse.

The only place where it could make sense is for tests, then then again, if you forget to call finish, your test will fail...